File: src/main/resources/static/resources/js/source/dto/migration-type.js
// @flow


export class MigrationType {

    /**
     * @type {MigrationType}
     */
    static FULL_MIGRATION : ?MigrationType = null;
    /**
     * @type {MigrationType}
     */
    static INCREMENTAL_MIGRATION : ?MigrationType = null;
    /**
     * @type {MigrationType}
     */
    static FULL_INCREMENTAL_MIGRATION : ?MigrationType = null;
    /**
     * @type {string}
     */
    name : string = '';

    /**
     * @param name {string}
     */
    constructor(name : string) {
        this.name = name;
    }

     toJSON() {
          return this.name;
     }
}

MigrationType.FULL_MIGRATION = new MigrationType('FULL_MIGRATION');
MigrationType.INCREMENTAL_MIGRATION = new MigrationType('INCREMENTAL_MIGRATION');
MigrationType.FULL_INCREMENTAL_MIGRATION = new MigrationType('FULL_INCREMENTAL_MIGRATION');

File: src/main/java/com/jhl/mds/consts/MigrationType.java
package com.jhl.mds.consts;

import lombok.Getter;

public enum MigrationType {

    FULL_MIGRATION(0b0001),
    INCREMENTAL_MIGRATION(0b0010),
    FULL_INCREMENTAL_MIGRATION(0b0011);

    @Getter
    private final int code;

    MigrationType(int code) {
        this.code = code;
    }

    public static MigrationType getByCode(int taskTypeCode) {
        for (MigrationType migrationType : MigrationType.values()) {
            if (migrationType.code == taskTypeCode) {
                return migrationType;
            }
        }
        throw new RuntimeException("Not found");
    }

    public boolean isApplicable(MigrationType migrationType) {
        return (this.code & migrationType.code) == this.code;
    }
}
